### PR TITLE
Fix CaptureGCHandleTargets taking forever to finish for large projects

### PR DESCRIPTION
**Pull request by Rohansi**
`CaptureGCHandleTargets` takes a very, very long time to complete in our project due to the improper use of a linked list. `g_list_append` doesn't look like it would be bad but it traverses to the end of the list *and then returns the starting node*. That's confusing, but it causes it to go through the entire list for every object.

I simply replaced the `GList` usage with a `GPtrArray` and can now actually take snapshots (30 minutes for a 10 GB one).

### CODE REVIEW

**Review by realek — APPROVED** [review]
LGTM, ty for the change :)

**Comment by Rohansi** [comment]
Is this still being worked on? I previously reported it as a bug as well (case 1142394) and was just wondering because we cannot take snapshots without using our own build of Mono.

**Comment by joncham** [comment]
@Rohansi sorry for delay, I am working with our legal to make sure all is in place for us to take contributions to this repo from non-Unity devs. Thanks for the change!

**Comment by joncham** [comment]
@Rohansi the bug you mentioned (1142394) has been marked as resolved. Is this PR still needed?

**Comment by Rohansi** [comment]
@joncham What was done to resolve it? I remember someone thinking that some of the issues I reported were duplicates because I reported both the capturing being excessively slow and the loading/processing being very slow.

This is related to capturing managed memory snapshots taking a very long time when dealing with very large heaps. I have personally used this change and it greatly sped up the capturing process, in fact I couldn't even take snapshots without this fix because they take absolutely forever.

**Comment by joncham** [comment]
@Rohansi can you rebase this change on latest `unity-master`, sign the CLA when prompted, and we can re-evaluate your contribution. Thanks!

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=1166) <br/>All committers have signed the CLA.

**Comment by Rohansi** [comment]
@joncham All done.

**Comment by Rohansi** [comment]
@joncham Any update on this?

**Comment by Rohansi** [comment]
*crickets*

**Comment by realek** [comment]
@joncham can we let this through? 📦 
Currently the memory profiler no longer uses any of the functionality that is being changed in this PR, as we exposed lower level logic from the VM to our backend. But quite a few people still do the [DllImport] for grabbing the capture_scripting_snapshot method and binding it in their C# snapshot constructs, building their own tools on top. As such this fix itself has value still ^_^, till we expose relevant APIs in C#

**Comment by joncham** [comment]
@realek I was just going to ping you. I am fine letting this through, just from my side I don't have any real validation that this is correct. So will merge if assuming you are good.

**Comment by realek** [comment]
> 
> 
> @realek I was just going to ping you. I am fine letting this through, just from my side I don't have any real validation that this is correct. So will merge if assuming you are good.

Tested it locally, it's all good ^^

**Pull request merged**
